packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -42,37 +43,38 @@ abstract class TextBoundary {
 
 /// A text boundary that uses characters as logical boundaries.
 ///
-/// This class takes grapheme clusters into account and avoid creating
-/// boundaries that generate malformed utf-16 characters.
+/// This class takes grapheme clusters into account and avoids creating
+/// boundaries that generate malformed UTF-16 characters.
 class CharacterBoundary extends TextBoundary {
   /// Creates a [CharacterBoundary] with the text.
-  const CharacterBoundary(this._text);
+  const CharacterBoundary(this._textEditingValue);
 
-  final String _text;
+  final TextEditingValue _textEditingValue;
```

Comment:
The goal for making public is to be reuse outside of EditableText, so it also needs to decouple TextEditingValue from TextBoundary class.

Is there a perf regression?

Reply:
BTW, what other widgets will be using the boundaries in the near future?
I'm thinking perhaps to improve the naming, we could have two classes:
- `TextValue` that simply holds a String (kind of acting like String view, although with a real String instance) - and use this as parms for CharacterBoundary, DocumentBoundary, _CodeUnitBoundary and _WhitespaceBoundary 
- have `TextEditingValue` extend `TextValue` by adding selection and composing

Reply:
SelectableRegion will reuse these classes. I think creating a delegate class to provide a typographic information sounds good. For now this delegate class will only have a string property, and we will likely need to add more as the feature expand. Also would like to hear suggestions from @LongCatIsLooong

Reply:
Thanks for looking into this! I wonder why there's copying, aren't strings supposed to be immutable?  @a-siva what's the recommended way to avoid the copying?

Reply:
The string should be passed by reference, so there should not be any copying. Are you sure that is the cost of CPU jump?